src/storage/hummock_sdk/src/slice_transform.rs

```diff
@@ -0,0 +1,423 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(dead_code)]
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use risingwave_common::catalog::ColumnDesc;
+use risingwave_common::types::VIRTUAL_NODE_SIZE;
+use risingwave_common::util::ordered::OrderedRowDeserializer;
+use risingwave_common::util::sort_util::OrderType;
+use risingwave_pb::catalog::Table;
+
+use crate::key::{get_table_id, TABLE_PREFIX_LEN};
+
+trait SliceTransform {
+    fn transform<'a>(&self, full_key: &'a [u8]) -> &'a [u8];
+}
+
+#[derive(Default)]
+pub struct FullKeySliceTransform;
+
+// Slice Transform generally used to transform key which will store in BloomFilter
+impl SliceTransform for FullKeySliceTransform {
+    fn transform<'a>(&self, full_key: &'a [u8]) -> &'a [u8] {
+        full_key
+    }
+}
+
+#[derive(Default)]
+pub struct DummySliceTransform;
+impl SliceTransform for DummySliceTransform {
+    fn transform<'a>(&self, full_key: &'a [u8]) -> &'a [u8] {
+        &full_key[0..0]
+    }
+}
+
+// SchemaSliceTransform build from table_catalog and transform a full_key to prefix for
+// prefix_bloom_filter
+pub struct SchemaSliceTransform {
+    // Each stateful operator has its own read pattern, partly using `Prefix Scan`.
+    // `PrefixKeyLength` can be decoded through its `DataType` and `OrderType` which obtained from
+    // `TableCatalog`. `read_pattern_prefix_column` means the count of column to decode prefix from
+    // `StorageKey`
+    read_pattern_prefix_column: u32,
+    deserializer: OrderedRowDeserializer,
+}
+
+impl SliceTransform for SchemaSliceTransform {
+    fn transform<'a>(&self, full_key: &'a [u8]) -> &'a [u8] {
+        let (_table_prefix, key) = full_key.split_at(TABLE_PREFIX_LEN);
+        let (_vnode_prefix, pk) = key.split_at(VIRTUAL_NODE_SIZE);
+
+        let pk_prefix_len = self
+            .deserializer
+            .deserialize_prefix_len_with_column_indices(
+                pk,
+                (0..self.read_pattern_prefix_column as usize).collect(),
+            )
+            .unwrap();
+
+        let prefix_len = TABLE_PREFIX_LEN + VIRTUAL_NODE_SIZE + pk_prefix_len;
```

Comment:
I'm not sure about this. Does `pk_prefix_len` also include join group key?

Reply:
yes, this pk its represents the StorageKey(like [table_id_vnode_pk_group_key_cid])